core/src/commonMain/kotlin/com/xebia/functional/xef/prompt/lang/Infer.kt

```diff
@@ -0,0 +1,114 @@
+ * LLM thought process. Assumes a LLM with Sudolang like capabilities or understanding of structured
+ * input.
+ */
+class Infer(
+  val model: ChatWithFunctions,
+  val conversation: Conversation,
+) {
+
+  class Replacement<out A>(val name: String, val value: A, val config: Config)
+
+  class Config(val modifiers: List<Pair<String, String>>) {
+    companion object {
+      operator fun invoke(vararg modifiers: Pair<String, String>): Config =
+        Config(modifiers.toList())
+    }
+  }
+
+  class Scope {
+
+    val replacements: MutableList<Replacement<*>> = mutableListOf()
+
+    val inferInt: Int = inferInt()
+    val inferFloat: Float = inferFloat()
+    val inferDouble: Double = inferDouble()
+    val inferString: String = inferString()
+
+    val placeholder
+      get() = "${'$'}generate${'$'}"
+
+    fun inferInt(config: Config = Config()): Int = infer(placeholder, Int.MAX_VALUE, config)
+
+    fun inferFloat(config: Config = Config()): Float = infer(placeholder, Float.MAX_VALUE, config)
+
+    fun inferDouble(config: Config = Config()): Double =
+      infer(placeholder, Double.MAX_VALUE, config)
+
+    fun inferString(config: Config = Config()): String = infer(placeholder, placeholder, config)
+
+    private fun <A> infer(name: String, value: A, config: Config): A {
+      replacements.add(Replacement(name, value, config))
+      return value
+    }
+  }
+
+  suspend inline operator fun <reified A, reified B> invoke(
+    prompt: Prompt,
+    block: Scope.() -> A
+  ): B {
+    val scope = Scope()
+    val request = block(scope)
+    var input = Json.encodeToString(serializer<A>(), request)
+    scope.replacements.forEach { replacement ->
+      input = replaceInferPlaceholder(input, replacement)
+    }
+    return model.prompt(
+      messages =
+        listOf(
+          Message.systemMessage { prompt.message },
+          Message.systemMessage {
+            "Stay in role and follow the directives of the function `Process`"
+          },
+          Message.systemMessage {
+            """
+          Process(input) {
+             STOP, Carefully consider all instructions in this function
+             STOP, first replace all placeholders ${scope.placeholder} in input 
+             Consider in your output ALL properties in the input that are not placeholders
+             Reflect their information in your output
+             target 
+              |> critique 
+              |> fix(critique) 
+              |> applyCritique(target)
+             produce output in valid json
+          }
+        """
+              .trimIndent()
+          },
+          Message.userMessage { "Process($input)" },
+        ),
+      scope = conversation,
+      serializer = serializer<B>(),
+      promptConfiguration = PromptConfiguration { temperature(0.0) }
+    )
+  }
+
+  fun replaceInferPlaceholder(input: String, replacement: Replacement<*>): String =
+    input.replaceFirst(
```

Comment:
We are replacing only the first occurrence, if we have 2 `inferInt` in the object, the first contains `$generate$` but the second contains the `Int.MAX_VALUE`

Do we want this?

Reply:
Finally, I have uploaded a new commit adding `get()` in order to generate all of the inferred data instead of one for every type